File: common/python_common/text_manipulate.py
# -*- coding: utf-8 -*-
#
#	text_manipulate.py
#
#					Aug/05/2014
import	sys
import	string
import	datetime
#
from to_utf8 import to_utf8_proc
# ---------------------------------------------------------------
def	text_read_proc	(file_in):
#
	fp_in = open (file_in,encoding='utf-8')
	lines = fp_in.readlines ()
	dict_aa = {}
	for line in lines:
		if (5 < len (line)):
			cols= line[:-1].split ('\t')
			if (3 < len (cols)):
				try:
					if (cols[0][0] == "t"):
						dict_unit = {'name': cols[1], \
			'population':int (cols[2]),'date_mod':cols[3]}
						dict_aa[cols[0]] = dict_unit
				except:
					sys.stderr.write \
				("*** error *** %s ***\n" % file_in)
					sys.stderr.write \
				("*** %s ***\n" % line)
#
	fp_in.close ()
	return	dict_aa
#
# ---------------------------------------------------------------
def	dict_display_proc (dict_aa):
	for key in sorted (dict_aa.keys()):
		if ((key != '_id') and (key != '_rev')):
			unit = dict_aa[key]
			name = unit['name']
			str_out = key+"\t"+ str (name)
			str_out += "\t" + str (unit['population'])
			str_out += "\t" + unit['date_mod']
			print	(str_out)
# ---------------------------------------------------------------
def	text_write_proc	(file_out,dict_aa):
#
	fp_out = open (file_out,mode='w', encoding='utf-8')
	for key in dict_aa.keys():
		unit = dict_aa[key]
		str_out = key + "\t" + str (unit['name']) + "\t"
		str_out += "%d\t" % unit['population']
		str_out += unit['date_mod'] + "\n"
		fp_out.write (str_out)
	fp_out.close ()
# ---------------------------------------------------------------
def	dict_update_proc (dict_in,id,population):
	key = str (id)
	if key in dict_in:
		dict_in[key]['population'] = population
		date_mod = datetime.date.today ()
		dict_in[key]['date_mod'] = '%s' % date_mod
#
	return	dict_in
#
# ---------------------------------------------------------------
def	dict_delete_proc (dict_in,key):
	if key in dict_in:
		del dict_in[key]
#
	return	dict_in
#
# ---------------------------------------------------------------
def	hash_update_proc (array_unit,population):
	date_mod = datetime.date.today ()
	array_unit['population'] = population
	array_unit['date_mod'] = '%s' % date_mod
# ---------------------------------------------------------------
def     dict_append_proc (dict_aa,key,name,population,date_mod):
	dict_aa[key] = {'name':name,'population':population,'date_mod':date_mod}
#	dict_aa[key] = {'name':name.encode ('utf-8'),'population':population,'date_mod':date_mod}
#
	return dict_aa
#
# ---------------------------------------------------------------
def	csv_write_proc	(file_out,dict_aa):
#
	fp_out = open (file_out,mode='w',encoding='utf-8')
	for key in dict_aa.keys():
		unit = dict_aa[key]
		str_out = key + "," + str (unit['name']) + ","
		str_out += "%d," % unit['population']
		str_out += unit['date_mod'] + "\n"
		fp_out.write (str_out)
#		fp_out.write (str_out.encode ('utf-8'))
	fp_out.close ()
# ---------------------------------------------------------------
def	csv_read_proc	(file_in):
#
	fp_in = open (file_in,encoding='utf-8')
	lines = fp_in.readlines ()
	dict_aa = {}
	for line in lines:
		cols= line[:-1].split (",")
		dict_unit = {'name': cols[1], \
			'population':int (cols[2]),'date_mod':cols[3]}
		dict_aa[cols[0]] = dict_unit
#
	fp_in.close ()
	return	dict_aa
#
# ---------------------------------------------------------------

File: common/ruby_common/text_manipulate.rb
#! /usr/bin/ruby
#
#	ruby_common/text_manipualate.rb
#
#					Oct/02/2015
#
# -------------------------------------------------------------
require	"date"
#
# -------------------------------------------------------------
def text_read_proc (file_in)
	ff = File.open(file_in,"r:UTF-8")
	dict_aa = {}
	while(line = ff.gets)
		line=line.chomp
		cols = line.split()
		if 2 < cols.length then
			unit = {}
#			unit['id'] = cols[0].to_i
			unit['name'] = cols[1]
			unit['population'] = cols[2].to_i
			unit['date_mod'] = cols[3]
			key = cols[0]
			dict_aa[key]=unit
		end
	end
	ff.close
	return	dict_aa
end
#

# -------------------------------------------------------------
def dict_display_proc (dict_aa)
	dict_aa.sort.each {|key, value|
		if (key.to_s != '_id') && (key.to_s != '_rev')
			print(key,"\t",value['name'],"\t")
			print(value['population'],"\t")
			print(value['date_mod'],"\n")
		end
	}
end
# -------------------------------------------------------------
def text_write_proc (file_out,dict_aa)
	puts "*** text_write_proc *** start ***"
	fp_out = File.open(file_out,"w")
	dict_aa.each {|key, value|
		fp_out.print key,"\t"
		fp_out.print value['name'],"\t"
		fp_out.print value['population'],"\t"
		fp_out.print value['date_mod'],"\n"
	}
	fp_out.close
	puts "*** text_write_proc *** end ***"
end
# -------------------------------------------------------------
def dict_update_proc (dict_aa,key,population_in)
	puts "*** dict_update_proc *** start *** " + key
	unit=dict_aa[key]
	print	unit['name'] + "\t"
	print	unit['population']
	puts
	unit['population'] = population_in
	unit['date_mod'] = Date.today
	puts "*** update_proc *** end ***"
#
	return	dict_aa
end

# -------------------------------------------------------------
def dict_delete_proc (dict_aa,key)
	puts "*** dict_delete_proc *** start ***"
	if (dict_aa.key?(key))
		unit=dict_aa[key]
		print	unit['name'],"\t"
		print	unit['population'],"\n"
		dict_aa.delete(key)
	end
	puts "*** delete_proc *** end ***"
#
	return	dict_aa
end

# -------------------------------------------------------------
def dict_append_proc (dict_aa,id,name,population,date_mod)
	unit = {}
	unit['name'] = name
	unit['population'] = population
	unit['date_mod'] = date_mod
	key = id.to_s
	dict_aa[key] = unit
	return dict_aa
end
# -------------------------------------------------------------
def csv_read_proc (file_in)
	ff = File.open(file_in,"r:UTF-8")
	dict_aa = {}
	while(line = ff.gets)
		line=line.chomp
		cols = line.split(",")
		unit = {}
		unit['name'] = cols[1]
		unit['population'] = cols[2].to_i
		unit['date_mod'] = cols[3]
		key = cols[0]
		dict_aa[key]=unit
	end
	ff.close
	return	dict_aa
end
#
# -------------------------------------------------------------
def csv_write_proc (file_out,dict_aa)
	puts "*** csv_write_proc *** start ***"
	fp_out = File.open(file_out,"w")
	dict_aa.each {|key, value|
		fp_out.print key,","
		fp_out.print value['name'],","
		fp_out.print value['population'],","
		fp_out.print value['date_mod'],"\n"
	}
	fp_out.close
	puts "*** csv_write_proc *** end ***"
end
# -------------------------------------------------------------
